### ISSUE TITLE
bizarre closure lossage

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9665]
Migrated from [rt.perl.org#63540](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=63540) (status was 'resolved')

Searchable as RT63540$

**Comment by p5pRT** [comment]
### From zefram@fysh.org

#### Created by zefram@fysh.org



$ cat t0\.pl
sub memoise \{
&nbsp;    my\($evaluate\) = @&#8203;\_;
&nbsp;    my $value;
&nbsp;    return sub \(\) \{
&nbsp;        \#if\(0\) \{ $evaluate; \}
&nbsp;        if\(defined $evaluate\) \{
&nbsp;            $value = $evaluate\->\(\);
&nbsp;            $evaluate = undef;
&nbsp;        \}
&nbsp;        return $value;
&nbsp;    \};
\}
my $a = memoise\(sub \{ "ok" \}\);
print $a\->\(\)\, "\\n";
$ perl5\.6\.0 t0\.pl
ok
$ perl5\.6\.2 t0\.pl
ok
$ perl5\.8\.0 t0\.pl

$ perl5\.8\.9 t0\.pl

$ perl5\.10\.0 t0\.pl
CODE\(0x8066ba8\)
$

5\.6 exhibits the expected behaviour\.  5\.8 and 5\.10 are giving nonsense
results\.  If the "if\(0\) \{ $evaluate; \}" line is uncommented\, which
should be doubly a no\-op\, then all of these perl versions give the
correct result\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.8:

Configured by Debian Project at Fri Dec 19 00:43:54 EST 2008.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.18-6-686, archname=i486-linux-gnu-thread-multi
    uname='linux etch 2.6.18-6-686 #1 smp fri dec 12 16:48:28 utc 2008 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.8 -Dsitearch=/usr/local/lib/perl/5.8.8 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.8 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.1.2 20061115 (prerelease) (Debian 4.1.1-21)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.6.so, so=so, useshrplib=true, libperl=libperl.so.5.8.8
    gnulibc_version='2.3.6'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /etc/perl
    /usr/local/lib/perl/5.8.8
    /usr/local/share/perl/5.8.8
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    /usr/local/lib/perl/5.8.4
    /usr/local/share/perl/5.8.4
    .


Environment for perl v5.8.8:
    HOME=/home/zefram
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=en_GB
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/zefram/pub/i686-pc-linux-gnu/bin:/home/zefram/pub/common/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/local/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Additional&#8203;: Matt Trout spotted that omitting the \(\) prototype from the
sub returned by memoise is another way to make it operate correctly on
all perl versions\.

I've also tried it on 5\.005 and 5\.004\, and they both operate correctly\.

\-zefram

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

On Thu Feb 26 12&#8203;:39&#8203;:26 2009\, zefram@&#8203;fysh\.org wrote&#8203;:
>
> $ cat t0\.pl
> sub memoise \{
>     my\($evaluate\) = @&#8203;\_;
>     my $value;
>     return sub \(\) \{
>         \#if\(0\) \{ $evaluate; \}
>         if\(defined $evaluate\) \{
>             $value = $evaluate\->\(\);
>             $evaluate = undef;
>         \}
>         return $value;
>     \};
> \}
> my $a = memoise\(sub \{ "ok" \}\);
> print $a\->\(\)\, "\\n";
> $ perl5\.6\.0 t0\.pl
> ok
> $ perl5\.6\.2 t0\.pl
> ok
> $ perl5\.8\.0 t0\.pl
> 
> $ perl5\.8\.9 t0\.pl
> 
> $ perl5\.10\.0 t0\.pl
> CODE\(0x8066ba8\)
> $
> 
> 5\.6 exhibits the expected behaviour\.  5\.8 and 5\.10 are giving 
nonsense
> results\.  If the "if\(0\) \{ $evaluate; \}" line is uncommented\, which
> should be doubly a no\-op\, then all of these perl versions give the
> correct result\.

Binary search&#8203;:

\-\-\-\-Program\-\-\-\-
sub memoise \{
&nbsp;        my\($evaluate\) = @&#8203;\_;
&nbsp;        my $value;
&nbsp;        return sub \(\) \{
&nbsp;                \#if\(0\) \{ $evaluate; \}
&nbsp;                if\(defined $evaluate\) \{
&nbsp;                        $value = $evaluate\->\(\);
&nbsp;                        $evaluate = undef;
&nbsp;                \}
&nbsp;                return $value;
&nbsp;        \};
\}
my $a = memoise\(sub \{ "ok" \}\);
print $a\->\(\)\, "\\n";
\_\_END\_\_


Change 1&#8203;:
\-\-\-\-Output of \.\.\.l/p1rFte9/perl\-5\.7\.0@&#8203;7386/bin/perl\-\-\-\-
ok

\-\-\-\-EOF \($?='0'\)\-\-\-\-
\-\-\-\-Output of \.\.\.l/ppS7IQi/perl\-5\.7\.0@&#8203;7389/bin/perl\-\-\-\-


\-\-\-\-EOF \($?='0'\)\-\-\-\-


http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/p/7389
Change 7389 by jhi@&#8203;alpha on 2000/10/21 14&#8203;:26&#8203;:45

&nbsp;    Subject&#8203;: Re&#8203;: Creating const subs for constants\.
&nbsp;    From&#8203;: John Tobey \<jtobey@&#8203;john\-edwin\-tobey\.org>
&nbsp;    Date&#8203;: Fri\, 20 Oct 2000 22&#8203;:03&#8203;:27 \-0400 \(EDT\)
&nbsp;    Message\-Id&#8203;: \<m13mo0N\-000FObC@&#8203;feynman\.localnet>

http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/p/7390
Change 7390 by jhi@&#8203;alpha on 2000/10/21 14&#8203;:35&#8203;:57

&nbsp;    Add a testcase for \#7389\.


Change 2&#8203;:
\-\-\-\-Output of \.\.\./pz1Wttf/perl\-5\.8\.0@&#8203;19635/bin/perl\-\-\-\-


\-\-\-\-EOF \($?='0'\)\-\-\-\-
\-\-\-\-Output of \.\.\./pMqzSN2/perl\-5\.8\.0@&#8203;19637/bin/perl\-\-\-\-
CODE\(0x8141478\)

\-\-\-\-EOF \($?='0'\)\-\-\-\-
Need a perl between 19635 and 19637


http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/p/19637
Change 19637 by rgs@&#8203;rgs\-home on 2003/05/29 18&#8203;:47&#8203;:40

&nbsp;    Subject&#8203;: \[PATCH\] jumbo closure fix
&nbsp;    From&#8203;: Dave Mitchell \<davem@&#8203;fdgroup\.com>
&nbsp;    Date&#8203;: Wed\, 26 Feb 2003 14&#8203;:49&#8203;:47 \+0000
&nbsp;    Message\-ID&#8203;: \<20030226144947\.A14444@&#8203;fdgroup\.com>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Mar 11\, 2009 at 05&#8203;:50&#8203;:54PM \-0700\, Bram via RT wrote&#8203;:
> On Thu Feb 26 12&#8203;:39&#8203;:26 2009\, zefram@&#8203;fysh\.org wrote&#8203;:
> >
> > $ cat t0\.pl
> > sub memoise \{
> >     my\($evaluate\) = @&#8203;\_;
> >     my $value;
> >     return sub \(\) \{
> >         \#if\(0\) \{ $evaluate; \}
> >         if\(defined $evaluate\) \{
> >             $value = $evaluate\->\(\);
> >             $evaluate = undef;
> >         \}
> >         return $value;
> >     \};
> > \}
> > my $a = memoise\(sub \{ "ok" \}\);
> > print $a\->\(\)\, "\\n";
> > $ perl5\.6\.0 t0\.pl
> > ok
> > $ perl5\.6\.2 t0\.pl
> > ok
> > $ perl5\.8\.0 t0\.pl
> > 
> > $ perl5\.8\.9 t0\.pl
> > 
> > $ perl5\.10\.0 t0\.pl
> > CODE\(0x8066ba8\)
> > $
> > 
> > 5\.6 exhibits the expected behaviour\.  5\.8 and 5\.10 are giving 
> nonsense
> > results\.  If the "if\(0\) \{ $evaluate; \}" line is uncommented\, which
> > should be doubly a no\-op\, then all of these perl versions give the
> > correct result\.
> 
> Binary search&#8203;:
> 
\[snip\]
> 
> http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/p/19637
> Change 19637 by rgs@&#8203;rgs\-home on 2003/05/29 18&#8203;:47&#8203;:40
> 
>     Subject&#8203;: \[PATCH\] jumbo closure fix
>     From&#8203;: Dave Mitchell \<davem@&#8203;fdgroup\.com>
>     Date&#8203;: Wed\, 26 Feb 2003 14&#8203;:49&#8203;:47 \+0000
>     Message\-ID&#8203;: \<20030226144947\.A14444@&#8203;fdgroup\.com>

There is a hacky optimisation that looks for code like this&#8203;:

&nbsp;    my $x;
&nbsp;    \.\.\.
&nbsp;    sub \(\) \{$x\}

ie an anon sub with \(\) prototype and whose body just returns an outer
lexical\, and if so\, optimises it into a CONST sub who's value is the
current $x\. This is needed to support 'use constant'\.

However\, the code in Perl\_op\_const\_sv\(\) seems to be failing to correctly
detect 'just a lexical' in the sub's body\.

In particular\, this fairly minimal example shows it returning a CONST sub&#8203;:

&nbsp;    use Devel&#8203;::Peek;
&nbsp;    my $x;
&nbsp;    Dump sub \(\) \{
&nbsp;        if \($x\) \{ \}
&nbsp;        return $x;
&nbsp;    \};

and its the second call to Perl\_op\_const\_sv\(\) that's getting it wrong\.
One day when I have more time\, I'll try to fix this\.


\-\- 
Thank God I'm an atheist\.\.\.\.\.

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Mar 23\, 2009 at 11&#8203;:39&#8203;:19AM \-0500\, David Nicol wrote&#8203;:
> On Sat\, Mar 21\, 2009 at 8&#8203;:44 AM\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> 
> > There is a hacky optimisation that looks for code like this&#8203;:
> >
> >    my $x;
> >    \.\.\.
> >    sub \(\) \{$x\}
> >
> > ie an anon sub with \(\) prototype and whose body just returns an outer
> > lexical\, and if so\, optimises it into a CONST sub who's value is the
> > current $x\. This is needed to support 'use constant'\.
> 
> What? That's horrible\!  How are we supposed to write read\-only
> accessors then? Oh\.  The hack is needed for
> http&#8203;://search\.cpan\.org/~saper/constant\-1\.17/lib/constant\.pm\#Magic\_constants
> to work\.  And read\-only accessors will still work as methods\, because
> methods aren't inlined\.

What are you jabbering on about? And what is the diff below apropos to?

> \-\-\- /usr/lib/perl5/5\.8/constant\_original\.pm     2009\-03\-23
> 10&#8203;:51&#8203;:16\.288169400 \-0500
> \+\+\+ /usr/lib/perl5/5\.8/constant\.pm      2009\-03\-23 11&#8203;:33&#8203;:18\.066527600 \-0500
> @&#8203;@&#8203; \-41\,12 \+41\,14 @&#8203;@&#8203;
>         $constants\{\+shift\} = undef;
>      \}
> 
> \+    my $pkg = caller;
> \+    my @&#8203;EvalMe = \("package $pkg;"\);
> \+
>      foreach my $name \( keys %constants \) \{
>         unless \(defined $name\) \{
>             require Carp;
>             Carp&#8203;::croak\("Can't use undef as constant name"\);
>         \}
> \-       my $pkg = caller;
> 
>         \# Normal constant name
>         if \($name =~ /^\_?\[^\\W\_0\-9\]\\w\*\\z/ and \!$forbidden\{$name\}\) \{
> @&#8203;@&#8203; \-94\,22 \+96\,37 @&#8203;@&#8203;
>             no strict 'refs';
>             my $full\_name = "$\{pkg\}&#8203;::$name";
>             $declared\{$full\_name\}\+\+;
> \+            my $body;
>             if \($multiple\) \{
>                 my $scalar = $constants\{$name\};
> \-               \*$full\_name = sub \(\) \{ $scalar \};
> \+                $scalar =~ /^\(\\d\*\\\.?\\d\+\)\\z/;
> \+               $body = \( $1 ? $1 : qq\{"\\Q$scalar\\E"\} \);
>             \} else \{
>                 if \(@&#8203;\_ == 1\) \{
>                     my $scalar = $\_\[0\];
> \-                   \*$full\_name = sub \(\) \{ $scalar \};
> \+                    $scalar =~ /^\(\\d\*\\\.?\\d\+\)\\z/;
> \+                    $body = \( $1 ? $1 : qq\{"\\Q$scalar\\E"\} \);
>                 \} elsif \(@&#8203;\_\) \{
>                     my @&#8203;list = @&#8203;\_;
> \-                   \*$full\_name = sub \(\) \{ @&#8203;list \};
> \+                   $body = join '\,'\, map \{
> \+                        /^\(\\d\*\\\.?\\d\+\)\\z/
> \+                        ? $1 : qq/"\\Q$\_\\E"/
> \+                    \} @&#8203;list;
>                 \} else \{
> \-                   \*$full\_name = sub \(\) \{ \};
> \+                    $body = '';
>                 \}
> \-           \}
> \+           \};
> \+            push @&#8203;EvalMe\, \<\<"THISPIECE";
> \+sub $name \(\) \{
> \+    $body
> \+\}
> \+THISPIECE
>         \}
>      \}
> \+    eval join "\\n"\,@&#8203;EvalMe\,"1;" or
> \+        die join "\\n"\,
> \+            "ERROR IN CODE CONSTRUCTION WITH"\,
> \+            @&#8203;EvalMe\,"END CODE"\,$@&#8203;;
>  \}
> 
>  1;
> @&#8203;@&#8203; \-270\,6 \+287\,9 @&#8203;@&#8203;
>  value\.  References to tied variables\, however\, can be used as
>  constants without any problems\.
> 
> \+This is the feature that is lost by using string\-eval rather than
> \+magical\, constantifying read\-only accessors\.
> \+
>  =head1 TECHNICAL NOTES
> 
>  In the current implementation\, scalar constants are actually

\-\- 
SCO \- a train crash in slow motion

**Comment by p5pRT** [comment]
### From @davidnicol

On Sat\, Mar 21\, 2009 at 8&#8203;:44 AM\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:

> There is a hacky optimisation that looks for code like this&#8203;:
>
>    my $x;
>    \.\.\.
>    sub \(\) \{$x\}
>
> ie an anon sub with \(\) prototype and whose body just returns an outer
> lexical\, and if so\, optimises it into a CONST sub who's value is the
> current $x\. This is needed to support 'use constant'\.

What? That's horrible\!  How are we supposed to write read\-only
accessors then? Oh\.  The hack is needed for
http&#8203;://search\.cpan\.org/~saper/constant\-1\.17/lib/constant\.pm\#Magic\_constants
to work\.  And read\-only accessors will still work as methods\, because
methods aren't inlined\.



\-\-\- /usr/lib/perl5/5\.8/constant\_original\.pm     2009\-03\-23
10&#8203;:51&#8203;:16\.288169400 \-0500
\+\+\+ /usr/lib/perl5/5\.8/constant\.pm      2009\-03\-23 11&#8203;:33&#8203;:18\.066527600 \-0500
@&#8203;@&#8203; \-41\,12 \+41\,14 @&#8203;@&#8203;
&nbsp;        $constants\{\+shift\} = undef;
&nbsp;     \}

\+    my $pkg = caller;
\+    my @&#8203;EvalMe = \("package $pkg;"\);
\+
&nbsp;     foreach my $name \( keys %constants \) \{
&nbsp;        unless \(defined $name\) \{
&nbsp;            require Carp;
&nbsp;            Carp&#8203;::croak\("Can't use undef as constant name"\);
&nbsp;        \}
\-       my $pkg = caller;

&nbsp;        \# Normal constant name
&nbsp;        if \($name =~ /^\_?\[^\\W\_0\-9\]\\w\*\\z/ and \!$forbidden\{$name\}\) \{
@&#8203;@&#8203; \-94\,22 \+96\,37 @&#8203;@&#8203;
&nbsp;            no strict 'refs';
&nbsp;            my $full\_name = "$\{pkg\}&#8203;::$name";
&nbsp;            $declared\{$full\_name\}\+\+;
\+            my $body;
&nbsp;            if \($multiple\) \{
&nbsp;                my $scalar = $constants\{$name\};
\-               \*$full\_name = sub \(\) \{ $scalar \};
\+                $scalar =~ /^\(\\d\*\\\.?\\d\+\)\\z/;
\+               $body = \( $1 ? $1 : qq\{"\\Q$scalar\\E"\} \);
&nbsp;            \} else \{
&nbsp;                if \(@&#8203;\_ == 1\) \{
&nbsp;                    my $scalar = $\_\[0\];
\-                   \*$full\_name = sub \(\) \{ $scalar \};
\+                    $scalar =~ /^\(\\d\*\\\.?\\d\+\)\\z/;
\+                    $body = \( $1 ? $1 : qq\{"\\Q$scalar\\E"\} \);
&nbsp;                \} elsif \(@&#8203;\_\) \{
&nbsp;                    my @&#8203;list = @&#8203;\_;
\-                   \*$full\_name = sub \(\) \{ @&#8203;list \};
\+                   $body = join '\,'\, map \{
\+                        /^\(\\d\*\\\.?\\d\+\)\\z/
\+                        ? $1 : qq/"\\Q$\_\\E"/
\+                    \} @&#8203;list;
&nbsp;                \} else \{
\-                   \*$full\_name = sub \(\) \{ \};
\+                    $body = '';
&nbsp;                \}
\-           \}
\+           \};
\+            push @&#8203;EvalMe\, \<\<"THISPIECE";
\+sub $name \(\) \{
\+    $body
\+\}
\+THISPIECE
&nbsp;        \}
&nbsp;     \}
\+    eval join "\\n"\,@&#8203;EvalMe\,"1;" or
\+        die join "\\n"\,
\+            "ERROR IN CODE CONSTRUCTION WITH"\,
\+            @&#8203;EvalMe\,"END CODE"\,$@&#8203;;
 \}

 1;
@&#8203;@&#8203; \-270\,6 \+287\,9 @&#8203;@&#8203;
 value\.  References to tied variables\, however\, can be used as
 constants without any problems\.

\+This is the feature that is lost by using string\-eval rather than
\+magical\, constantifying read\-only accessors\.
\+
 =head1 TECHNICAL NOTES

 In the current implementation\, scalar constants are actually

**Comment by p5pRT** [comment]
### From @nwc10

On Sat\, Mar 21\, 2009 at 01&#8203;:44&#8203;:40PM \+0000\, Dave Mitchell wrote&#8203;:

> There is a hacky optimisation that looks for code like this&#8203;:
> 
>     my $x;
>     \.\.\.
>     sub \(\) \{$x\}
> 
> ie an anon sub with \(\) prototype and whose body just returns an outer
> lexical\, and if so\, optimises it into a CONST sub who's value is the
> current $x\. This is needed to support 'use constant'\.
> 
> However\, the code in Perl\_op\_const\_sv\(\) seems to be failing to correctly
> detect 'just a lexical' in the sub's body\.
> 
> In particular\, this fairly minimal example shows it returning a CONST sub&#8203;:
> 
>     use Devel&#8203;::Peek;
>     my $x;
>     Dump sub \(\) \{
>         if \($x\) \{ \}
>         return $x;
>     \};
> 
> and its the second call to Perl\_op\_const\_sv\(\) that's getting it wrong\.
> One day when I have more time\, I'll try to fix this\.

If anyone wants to beat Dave to it\, the code in question starts here
http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/HEAD&#8203;:/op\.c\#l5401

Nicholas Clark

**Comment by p5pRT** [comment]
### From @davidnicol

On Mon\, Mar 23\, 2009 at 1&#8203;:43 PM\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> What are you jabbering on about? And what is the diff below apropos to?

I rewrote constant\.pm to create a code block to eval\-string\, instead
of assigning anonsubs to globs\.

After reading Nick's last\, I see that the thing that needs to happen
is that the constify hack should only happen when there are only so
many references to the lexical\, so


&nbsp;  BEGIN \{
&nbsp;    my \($x\, $y\, $z\) = \($$\, "undefined"\,$$\);
&nbsp;    \*const63540 = sub \(\) \{ $x \};
&nbsp;    \*pidconst63540 = sub \(\) \{ $z \};
&nbsp;    \*getthing63540 = sub \(\) \{ $y \};
&nbsp;    \*setthing63540 = sub \($\) \{ $y = shift \};
&nbsp;    $x = 63540
&nbsp;  \}

will then optimize pidconst but not getthing\, which could be tested
with the above and then

&nbsp;  ok \(63540 == const63540\);
&nbsp;  ok \($$ == pidconst63540\);
&nbsp;  setthing63540 "tintinabulation of the bells bells bells";
&nbsp;  ok \(getthing63540 eq "tintinabulation of the bells bells bells"\);

Hmm\.  It isn't the reference counts\, we're checking for that correctly

<details><summary>Inline Patch</summary>

```diff
diff --git a/op.c b/op.c
index 78d9990..d0105ad 100644
--- a/op.c
+++ b/op.c
@@ -5460,6 +5460,7 @@ Perl_op_const_sv(pTHX_ const OP *o, CV *cv)
                sv = PAD_BASE_SV(CvPADLIST(cv), o->op_targ);
                /* the candidate should have 1 ref from this pad and 1 ref
                 * from the parent */
+                warn ("sv %s has refcount %i", SvPV_nolen(sv),

```
</details>
SvREFCNT(sv)); /* dln */
&nbsp;                if \(\!sv || SvREFCNT\(sv\) \!= 2\)
&nbsp;                    return NULL;
&nbsp;                sv = newSVsv\(sv\);


and by reordering things

&nbsp;  BEGIN \{
&nbsp;    my \($x\, $y\, $z\) = \($$\, "undefined"\,$$\);
&nbsp;    \*setthing63540 = sub \($\) \{ $y = shift \};
&nbsp;    $x = 63540;
&nbsp;    \*const63540 = sub \(\) \{ $x \};
&nbsp;    \*pidconst63540 = sub \(\) \{ $z \};
&nbsp;    \*getthing63540 = sub \(\) \{ $y \};
&nbsp;  \}

so that $x is set before creating the sub\, and the other reference is
made before creating the sub\, the tests pass\.

Deferring the constification of the sub until the routine is used
instead of when it is assigned to the glob \(if possible\)  would fix
both cases\, as by then $x would be set and $y would have a higher
refcount\.

Document this as yet another thing that you just have to be aware of?
Define setters before getters to avoid triggering constanting?


\-\- 
"I lived in Hannibal fifteen and a half years\, altogether\, then ran
away\, according to the custom of persons who are intending to become
celebrated\." \-\- Samuel Clemens

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Mar 24\, 2009 at 04&#8203;:15&#8203;:52PM \-0500\, David Nicol wrote&#8203;:

> After reading Nick's last\, I see that the thing that needs to happen
> is that the constify hack should only happen when there are only so
> many references to the lexical\, so

I don't think that this is right\.

> Deferring the constification of the sub until the routine is used
> instead of when it is assigned to the glob \(if possible\)  would fix
> both cases\, as by then $x would be set and $y would have a higher
> refcount\.
> 
> Document this as yet another thing that you just have to be aware of?
> Define setters before getters to avoid triggering constanting?

I think that the correct fix is to make Perl\_cv\_const\_sv\(\) do what it's
intended to do\, rather than changing any of the code in Perl space\. I like
the intended behaviour \- that a closure with a prototype of \(\) that returns
only a constant behaves like a subroutine with a prototype of \(\) that returns
only a constant \- both being marked as inlinable by the tokeniser\.

Making other changes would fix the symptoms\, rather than the bug\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Mar 24\, 2009 at 09&#8203;:48&#8203;:58PM \+0000\, Nicholas Clark wrote&#8203;:
> On Tue\, Mar 24\, 2009 at 04&#8203;:15&#8203;:52PM \-0500\, David Nicol wrote&#8203;:
> I think that the correct fix is to make Perl\_cv\_const\_sv\(\) do what it's
> intended to do\, rather than changing any of the code in Perl space\. I like
> the intended behaviour \- that a closure with a prototype of \(\) that returns
> only a constant behaves like a subroutine with a prototype of \(\) that returns
> only a constant \- both being marked as inlinable by the tokeniser\.
> 
> Making other changes would fix the symptoms\, rather than the bug\.

Exactly\!

\-\- 
"You may not work around any technical limitations in the software"
&nbsp;    \-\- Windows Vista license

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Mar 24\, 2009 at 04&#8203;:15&#8203;:52PM \-0500\, David Nicol wrote&#8203;:
> On Mon\, Mar 23\, 2009 at 1&#8203;:43 PM\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> > What are you jabbering on about? And what is the diff below apropos to?
> 
> I rewrote constant\.pm to create a code block to eval\-string\, instead
> of assigning anonsubs to globs\.

I could see what the patch did\. I just didn't know \*why\* you had submitted
the patch\. I didn't know know whether it was intended to be a serious
proposal for a fix to be committed\, or a proof of concept to illustrate
something\, or an irrelevant "ooh look at me\, aren't I clever" post\, or
whatever\.

Because you didn't bother to say\.

So I have to waste my fucking time on it \(when I know with 99% certainty
that in the end I'll discover it was bollocks anyway\)\.

PS \- if you'd read my original post properly\, you'd have seen that I'd
already diagnosed the problem and even identified the function at fault\,
and in what way it was at fault\.

\-\- 
In England there is a special word which means the last sunshine
of the summer\. That word is "spring"\.

**Comment by p5pRT** [comment]
### From @davidnicol

I have been asked what the patch I submitted on march 23 was supposed
to be or do\, so here is an explanation\.

I had thought that the constant pragma created some strings and then
evaluated them\, which as I understood things was the userland way of
creating a SvCONST subroutine\.  It is\, except that magic variables\,
which are supported by use constant\, are not supported by that method\.

The included patch is what it takes to switch C\<use constant> from its
current method to a method involving evaluating constructed strings\.
It was intended for discussion\, not inclusion\.

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed by dbe92b04c\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Nov 29 06&#8203;:10&#8203;:58 2010\, sprout wrote&#8203;:
> Fixed by dbe92b04c\.

But that fix papered over the inherent problem with the code&#8203;:  We are following op\_next pointers without starting at the beginning of the chain\.  I was trying to understand the logic in op\_const\_sv this morning when it dawned on me that there wasn’t any\.  I’m about to refactor op\_const\_sv into 3 subs \(it does three different things anyway\)\.  Then things will start to make sense and this type of bug will be less likely to recur\.

\-\- 

Father Chrysostomos
